lib/jekyll/filters/grouping_filters.rb

```diff
@@ -0,0 +1,63 @@
+module Jekyll
+  module Filters
+    module GroupingFilters
+      # Group an array of items by a property
+      #
+      # input - the inputted Enumerable
+      # property - the property
+      #
+      # Returns an array of Hashes, each looking something like this:
+      #  {"name"  => "larry"
+      #   "items" => [...] } # all the items where `property` == "larry"
+      def group_by(input, property)
+        if groupable?(input)
+          groups = input.group_by { |item| item_property(item, property).to_s }
+          make_grouped_array(groups)
+        else
+          input
+        end
+      end
+
+      # Group an array of items by an expression
+      #
+      # input - the object array
+      # variable - the variable to assign each item to in the expression
+      # expression -a Liquid comparison expression passed in as a string
+      #
+      # Returns the filtered array of objects
+      def group_by_exp(input, variable, expression)
+        return input unless groupable?(input)
+
+        parsed_expr = parse_expression(expression)
+        @context.stack do
+          groups = input.group_by do |item|
+            @context[variable] = item
+            parsed_expr.render(@context)
+          end
+          make_grouped_array(groups)
+        end
+      end
+
+      private
+      def parse_expression(str)
+        Liquid::Variable.new(str, {})
+      end
+
+      private
+      def groupable?(element)
+        element.respond_to?(:group_by)
+      end
+
+      private
+      def make_grouped_array(groups)
```

Comment:
I don't like the name of this method – something about Ruby conventions which I always loved is that the method could be considered like a variable, so you name things as though they *aren't* dynamic. In this case, you'd remove `make_` and just have `grouped_array(groups)` or `group_metadata_array(groups)`. Would either of those suit you?